src/coreclr/jit/assertionprop.cpp

```diff
@@ -3177,6 +3177,67 @@ GenTree* Compiler::optVNConstantPropOnTree(BasicBlock* block, GenTree* tree)
 
     if (conValTree != nullptr)
     {
+        if (tree->OperIs(GT_LCL_VAR) && conValTree->OperIs(GT_CNS_VEC, GT_CNS_DBL))
+        {
+            // A simple heuristic: If the constant is defined outside of a loop (not far from its head)
+            // and is used inside it - don't propagate.
+
+            bool keepPropagating = false;
+
+            if (conValTree->OperIs(GT_CNS_VEC))
```

Comment:
Is the bespoke profitability analysis necessary -- could we leverage `gtSetEvalOrder` (aka `gtPrepareCost`) here?